nativelib/src/main/scala/java/lang/ieee754tostring/ryu/RyuDouble.scala

```diff
@@ -0,0 +1,548 @@
+// Copyright 2018 Ulf Adams
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Credits:
+//
+//    1) This work is a heavily modified derivation of the original work by
+//       Ulf Adams at URL: https://github.com/ulfjack/ryu.
+//       As such, it inherits the Apache license of the original work.
+//       Thank you Ulf Adams.
+//
+//    2) The original java sources were converted to a rough draft of
+//       scala code using the service at URL: javatoscala.com.
+//
+//       The raw conversion did not compile and contained bugs due
+//       to the handling of break and return statements, but it saved
+//       days, if not weeks, of effort.
+//
+//       Thank you javatoscala.com.
+//
+//    3) All additional work, including introduced bugs,  is an original
+//       contribution to Scala Native development.
+
+package java.lang.ieee754tostring.ryu
+
+import java.math.BigInteger
+
+import RyuRoundingMode._
+
+object RyuDouble {
+
+  var DEBUG: Boolean = false
+
+  private val DOUBLE_MANTISSA_BITS: Int = 52
+
+  private val DOUBLE_MANTISSA_MASK: Long = (1L << DOUBLE_MANTISSA_BITS) - 1
+
+  private val DOUBLE_EXPONENT_BITS: Int = 11
+
+  private val DOUBLE_EXPONENT_MASK: Int = (1 << DOUBLE_EXPONENT_BITS) - 1
+
+  private val DOUBLE_EXPONENT_BIAS: Int = (1 << (DOUBLE_EXPONENT_BITS - 1)) - 1
+
+  private val POS_TABLE_SIZE: Int = 326
+
+  private val NEG_TABLE_SIZE: Int = 291
+
+  // Only for debugging.
+  private val POW5: Array[BigInteger] = new Array[BigInteger](POS_TABLE_SIZE)
+
+  private val POW5_INV: Array[BigInteger] =
+    new Array[BigInteger](NEG_TABLE_SIZE)
+
+  private val POW5_BITCOUNT: Int = 121 // max 3*31 = 124
+
+  private val POW5_QUARTER_BITCOUNT: Int = 31
+
+  private val POW5_SPLIT: Array[Array[Int]] =
+    Array.ofDim[Int](POS_TABLE_SIZE, 4)
+
+  private val POW5_INV_BITCOUNT: Int = 122 // max 3*31 = 124
+
+  private val POW5_INV_QUARTER_BITCOUNT: Int = 31
+
+  private val POW5_INV_SPLIT: Array[Array[Int]] =
+    Array.ofDim[Int](NEG_TABLE_SIZE, 4)
+
+  val mask: BigInteger = BigInteger
```

Comment:
BigIntegers are really big in terms of implementation code size (i.e. native binary size too). We should really remove dependency on them in our implementation.

Reply:
Understood. I'll check what the Ryu C code does & check back here (probably
take a few days to study  and breadboard, especially if I have to dive into
codegen.).

IIRC,  The SN port of Harmony  <mumble>Format <mumble> uses BigInteger
all over the place (and does a  lot of hideous manipulation.   This PR gets
1.0.toString and Double.toString to work, but "%s".format(1.0) will still bring
in all the BigInteger stuff.  Again, from memory of pain. the "%s" form is
what is used by printf and string interpolation s"".  Not an excuse to 
let BigInteger creep in here if we can avoid it. 

Until the next leg of the ultra-marathon relay
HelloWorld.scala using Double.toString may avoid BigInterger but
 HelloWorld.scala doing a printf of a Double or Float will bring it in.

The galaxy sized issues are related. This PR identifies both but
focuses on solving  Double.toString. 

The size of the data structures in Ryu concern me but they are almost
certainly smaller then the Harmony BigInteger handling.  I never could
get the Harmony Double code to generate correct strings, especially
in the least significant digit or three.

Reply:
Yep, formatting is also another area that needs work to reduce binary size. To best of my knowledge we don't depend on it by default which is a good thing. Lets try to avoid the dependency here as {Float, Double}.toString are always linked by default (they are actually reachable).

Reply:
@densh  BigInteger is now gone from RyuDouble.scala and the resultant passes test-all.
If/When those changes pass review, I can make similar changes to RyuFloat.scala.

Let me point out two concerns for your review:

* I used scala.Array() because it allowed me to initialize the Array & because I am used to it.
  Does that have some performance impact?  Should I be doing something with
  scalanative.runtime.IntArray()? I believe that the latter is used under scala.Array().
  How heavy is the overhead? 

* Is the size of the 2D arrays POW5_SPLIT & POW5_INV_SPLIT acceptable?
  They will increase the byte count for simple HelloWorld programs and I know that
   would are working to minimize that.

Thank you.

Reply:
1. `scala.Array[Int]` and `scalanative.runtime.IntArray` are effectively the same thing (one compiled down to another). You should use `scala.Array` here.

2. Can we use one-dimensional array instead? It would be more compact in compiled code. For example, instead of:
```
val array = Array(
  Array(1, 2, 3, 4),
  Array(5, 6, 7, 8),
  ...
)
```
Use:
```
val array = Array(
  1, 2, 3, 4,
  5, 6, 7, 8,
  ...
)
```
And then for indexing, instead of `array(i)(j)` you can do `array(i * 4 + j)`. One-dimensional int arrays of constant values are relatively compact in binary form (4 bytes per element).

Reply:
Another thing it would be good to leave a comment that explains how this constant array was obtained. A snippet of code in the comment that generates a text of this array would be the most obvious documentation.

Reply:
Done!  I feel more confident knowing that scala.Array is the one to use here.

The 1D array indexing array(i * 4 + j) is a row major reminder of my Fortran computer
graphics days. Very familiar.

All BigInteger code is now removed from Ryu*.scala.  As requested, I also added a
longish comment describing how the new array constants were derived. In case somebody
in the needs to re-generate or fix the values.  Unfortunately, almost all the work was
done by hand. 

The review evolutions have, in my opinion, made the code __vastly__ better.